feat(core): support TypeScript 5.3

crisbeto opened the pull request:
Updates the repo to support TypeScript 5.3 and resolve any issues. Fixes include:
* Updating usages of TS compiler APIs to match their new signatures.
* In TS 5.3 negative numbers are represented as `PrefixUnaryExpression` instead of `NumericExpression`. These changes update all usages to account for it since passing a negative number into the old APIs results in a runtime error.

JoostK commented on integration/typings_test_ts53/package.json:
```diff
@@ -0,0 +1,29 @@
+{
+  "name": "angular-integration",
+  "description": "Assert that users with TypeScript 5.2 can type-check an Angular application",
```
```suggestion
  "description": "Assert that users with TypeScript 5.3 can type-check an Angular application",
```

JoostK commented on goldens/public-api/localize/tools/index.md:
````diff
@@ -4,6 +4,10 @@
 
 ```ts
 
+/// <reference types="@angular/compiler-cli/private/babel" />
+/// <reference types="@types/babel__core" />
+/// <reference types="@types/babel__traverse" />
````
Are these also emitted into the `.d.ts` files that are shipped to NPM? If so, would the reference directives require those modules to be present in a user's node_modules tree? The `@types/babel__*` packages aren't deps of localize, so they may not be available.

crisbeto replied:
Doesn't look like it. Here's what the `dist/packages-dist/localize/tools/index.d.ts` looks like when it's built with TS 5.3:

```
export { DiagnosticHandlingStrategy, Diagnostics } from './src/diagnostics';
export { checkDuplicateMessages } from './src/extract/duplicates';
export { MessageExtractor } from './src/extract/extraction';
export { ArbTranslationSerializer } from './src/extract/translation_files/arb_translation_serializer';
export { SimpleJsonTranslationSerializer } from './src/extract/translation_files/json_translation_serializer';
export { LegacyMessageIdMigrationSerializer } from './src/extract/translation_files/legacy_message_id_migration_serializer';
export { Xliff1TranslationSerializer } from './src/extract/translation_files/xliff1_translation_serializer';
export { Xliff2TranslationSerializer } from './src/extract/translation_files/xliff2_translation_serializer';
export { XmbTranslationSerializer } from './src/extract/translation_files/xmb_translation_serializer';
export { buildLocalizeReplacement, isGlobalIdentifier, translate, unwrapExpressionsFromTemplateLiteral, unwrapMessagePartsFromLocalizeCall, unwrapMessagePartsFromTemplateLiteral, unwrapSubstitutionsFromLocalizeCall } from './src/source_file_utils';
export { makeEs2015TranslatePlugin } from './src/translate/source_files/es2015_translate_plugin';
export { makeEs5TranslatePlugin } from './src/translate/source_files/es5_translate_plugin';
export { makeLocalePlugin } from './src/translate/source_files/locale_plugin';
export { ArbTranslationParser } from './src/translate/translation_files/translation_parsers/arb_translation_parser';
export { SimpleJsonTranslationParser } from './src/translate/translation_files/translation_parsers/simple_json_translation_parser';
export { Xliff1TranslationParser } from './src/translate/translation_files/translation_parsers/xliff1_translation_parser';
export { Xliff2TranslationParser } from './src/translate/translation_files/translation_parsers/xliff2_translation_parser';
export { XtbTranslationParser } from './src/translate/translation_files/translation_parsers/xtb_translation_parser';
```

crisbeto replied:
Actually nevermind, they do get emitted into some of those re-exported files. @devversion should we list these packages as dependencies?

devversion replied:
Even before this update, this dependency already existed (because `@babel/core` is imported in the `.d.ts`). In addition there is also the ESM explicit extension issue with `localize/tools` (something on the backburner).

I think we can safely add the type dependency, in case some people check the types (`skipLibCheck  = false`). I wonder if someone has a reference to the TS commit that changed this behavior.

dgp1130 replied:
Neither of these packages expose global types right? If so, then I think it should be safe to include the dependencies. I'm not very familiar with with `<reference type="..." />` or how Babel interacts with `@angular/localize` though. @clydin, do you have any concerns with this?

JoostK commented on packages/compiler-cli/src/ngtsc/translator/src/type_emitter.ts:
```diff
@@ -114,31 +114,15 @@ export class TypeEmitter {
         if (ts.isTypeReferenceNode(node)) {
           return this.emitTypeReference(node);
+        } else if (
+            // As of TypeScript 5.3 negative numbers are `PrefixUnaryExpression`
+            // instead of `NumericLiteral`.
+            ts.isPrefixUnaryExpression(node) && node.operator === ts.SyntaxKind.MinusToken &&
+            ts.isNumericLiteral(node.operand)) {
+          return ts.factory.createPrefixUnaryExpression(
+              node.operator, this.cloneLiteral(node.operand));
```
SInce all nodes are visited recursively using `visitNode`, do we really need to special case this? I'd expect the recursion to end up at the bare literal within the prefix unary expression and clone it at that point.

crisbeto committed:
feat(core): support TypeScript 5.3

Updates the repo to support TypeScript 5.3 and resolve any issues. Fixes include:
* Updating usages of TS compiler APIs to match their new signatures.
* In TS 5.3 negative numbers are represented as `PrefixUnaryExpression` instead of `NumericExpression`. These changes update all usages to account for it since passing a negative number into the old APIs results in a runtime error.

crisbeto commented:
I've updated `localize` to have dependencies on `@types/babel__core` and `@types/babel__traverse`.

josephperrott reviewed (approved):
LGTM

atscott reviewed (approved):
reviewed-for: public-api

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

jessicajaniuk commented:
This PR was merged into the repository by commit 94096c6ede67436a349ae07901f2bb418bf9f461.
